src/ws/session-utils.c

```diff
@@ -399,11 +399,61 @@ open_session (pam_handle_t *pamh)
   return PAM_SUCCESS;
 }
 
+static void
+do_fd_remapping (int *remap_fds, int n_remap_fds)
+{
+  /* we need to get all of the remap-fds to be numerically above
+   * n_remap_fds in order to make sure that we don't overwrite them in
+   * the middle of the dup2() loop below, and also avoid the case that
+   * dup2() is a no-op (which could fail to clear the O_CLOEXEC flag,
+   * for example).
+   *
+   * we're sloppy with the old fd values because we know that the fdwalk
+   * below will clean everything up.
+   */
+  for (int i = 0; i < n_remap_fds; i++)
+    if (remap_fds[i] != -1) /* -1 == no-op */
+      {
+        /* keep trying until we get a large enough fd.
+         *
+         * it might already be large enough, in which case we don't have
+         * to do anything.
+         */
+        while (remap_fds[i] < n_remap_fds)
+          {
+            int new_fd = dup (remap_fds[i]);
+
+            if (new_fd == -1)
+              {
+                warn ("dup(%d) failed", remap_fds[i]);
+                _exit (42);
```

Comment:
OOI, is `err()`/`errx()` not safe after forking?

Reply:
To be honest, we're doing an incredibly large amount of "unsafe after forking stuff" here.  We're calling Kerberos library functions...  The implementation of `fdwalk()` is also unsafe.  I think this is the least of our concerns.

Reply:
Not sure what you meant, but *I* meant that `warn()` + `_exit()` sounds a lot like `err()`, except that the latter may call `exit()` instead of `_exit()`. I was just just curious why you didn't use `err()` here.